habitat-baselines/habitat_baselines/articulated_agent_controllers/humanoid_rearrange_controller.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Meta Platforms, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import os
+import pickle as pkl
+from typing import List
+
+import magnum as mn
+import numpy as np
+
+
+class Pose:
+    def __init__(self, joints, obj_transform):
+        """
+        Contains a single humanoid pose
+            :param joints: list or array of num_joints * 4 elements, with the rotation quaternions
+            :pram obj_transform: Matrix4 with the root trasnform.
+        """
+        self.joints = list(joints)
+        self.root_transform = obj_transform
+
+
+class Motion:
+    """
+    Contains a sequential motion, corresponding to a sequence of poses
+        :param pose_array: num_poses x num_joints x 4 array, containing the join orientations
+        :param transform_array: num_poses x 4 x 4 array, containing the root transform
+        :param displacement: on each pose, how much forward displacement was there?
+            Used to measure how many poses we should advance to move a cerain amount
+        :param fps: the FPS at which the motion was recorded
+    """
+
+    def __init__(self, pose_array, transform_array, displacement, fps):
+        num_poses = pose_array.shape[0]
+        self.num_poses = num_poses
+        poses = []
+        for index in range(num_poses):
+            pose = Pose(
+                pose_array[index].reshape(-1),
+                mn.Matrix4(transform_array[index]),
+            )
+            poses.append(pose)
+
+        self.poses = poses
+        self.fps = fps
+        self.displacement = displacement
+
+
+MIN_ANGLE_TURN = 0  # If we turn less than this amount, we can just rotate and walk as if we had not rotated
+TURNING_STEP_AMOUNT = 20  # The maximum we should be turning at a time
```

Comment:
is this an angle? What is "at a time" in relation to "num_poses"?

Reply:
Yes, I replaced it for at a given step.